### PR TITLE
add tools/depcheck README

### PR DESCRIPTION
**Pull request by juanvallejo**
~~Work in progress...~~

~~Opening now to collect feedback as this is being written.~~

cc @deads2k @soltysh

### CODE REVIEW

File: tools/depcheck/README.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+### Getting started
+
+#### Limitations
+
+Before beginning to use this tool, it is important to note that because both the `trace`
+and `analyze` commands make use of [go list](https://golang.org/cmd/go/#hdr-List_packages)
+in order to obtain information about packages in a Go repository, there are limitations
+that exist for these two commands:  
+
+  - No symlinked directories will be traversed in a given repo. Any resulting graph information generated
```
Comment (soltysh) [inline tools/depcheck/README.md]: Since we heavily rely on this I'd like to see a copy&paste command how to solve this for origin.

File: tools/depcheck/README.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+by these commands will simply not include these directories or their subtrees.
+
+  - A `GOPATH` must be set prior to their use. This allows each command to deal with paths
```
Comment (soltysh) [inline tools/depcheck/README.md]: Not quite sure I follow what you mean with that?

File: tools/depcheck/README.md (hunk 1 above)
Comment (soltysh) [reply tools/depcheck/README.md]: Ok, since I'm commenting as I'm reading I've got to a place where you do so, can you link it here?

File: tools/depcheck/README.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+Finally, compile the `depcheck` binary:
+
+```bash
+$ go build /path/to/openshift/origin/tools/depcheck/depcheck.go
+```
+
+After these steps, you should have a `depcheck` binary, and a Go environment suitable for
+running the `trace` and `analyze` commands. 
+
+### Filters
```
Comment (soltysh) [inline tools/depcheck/README.md]: I built it, but how do I run it? A running section is needed, and only after that explain filters and excludes. I got lost jumping from compilation straight into filters.

File: tools/depcheck/README.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
+depcheck
+========
+
+Dependency analysis tool for Go repositories.
```
Comment (soltysh) [inline tools/depcheck/README.md]: Add a tl;dr section at the top.

File: tools/depcheck/README.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+### Why
```
Comment (soltysh) [inline tools/depcheck/README.md]: `## Why`

File: tools/depcheck/README.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+$ mkdir -p /tmp/depcheck/src/github.com/openshift
+$ cp -rL /path/to/openshift/origin /tmp/depcheck/src/github.com/openshift/origin
+
+# set the GOPATH to look in the new temporary location, in order
+# to ensure we traverse packages of this particular directory, and
+# not the original "origin" directory
+$ export GOPATH=/tmp/depcheck
+```
+
+#### Compiling
```
Comment (soltysh) [inline tools/depcheck/README.md]: Move the compilation and the running section right after `Getting started`.

File: tools/depcheck/README.md (hunk 5 above)
Comment (soltysh) [reply tools/depcheck/README.md]: Generally I'd like to see the following structure:

TL;DR
1. Why
2. Getting started
  * Compiling
  * Running (put examples here)
  * Limitations
3. Advanced usage
  * Analyze (put examples here)
  * Trace (put examples here)
  * Filters
  * Excludes

File: tools/depcheck/README.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  - `--filter=`   The path to the `filters.json` file containing package paths to "squash".
+  - `--exclude=`  The path to the `excludes.json` file containing package paths to exclude from the internal graph and analysis.
+  - `--openshift` Mutually exclusive with `--filter` and `--exclude`. Uses `openshift` specific repository information to generate an internal list of package paths to exclude and filter (requires that the `origin` repository exist within your `GOPATH`).
+
+Example:
+
+Invoke the `analyze` command using `openshift`-specific defaults, comparing dependencies against those of `kubernetes`:
+
+```bash
+$ depcheck analyze --root=github.com/openshift/origin --entry=cmd/... --entry=pkg/... --entry=tools/... --entry=test/... --openshift --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
```
Comment (soltysh) [inline tools/depcheck/README.md]: I created the tmp gopath exported `$GOPATH` and run this command insinde `tools/godepcheck` (because that's where the binary is) and in origin root and I got:
```
$ ./depcheck analyze --root=github.com/openshift/origin --entry=cmd/... --entry=pkg/... --entry=tools/... --entry=test/... --openshift --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
can't load package: package github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings: code in directory /tmp/depcheck/src/github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings expects import "github.com/kubernetes-incubator/service-catalog/pkg/apis/settings"
can't load package: package github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/v1alpha1: code in directory /tmp/depcheck/src/github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/v1alpha1 expects import "github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/v1alpha1"
can't load package: package github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/registry/settings/podpreset: code in directory /tmp/depcheck/src/github.com/openshift/origin/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/pkg/registry/settings/podpreset expects import "github.com/kubernetes-incubator/service-catalog/pkg/registry/settings/podpreset"
Error: exit status 1
Usage:
 ...
exit status 1
```

Btw. with error don't print all the usage, unless I specify wrong flag, similarly how we do it with `oc`.

File: tools/depcheck/README.md (hunk 1 above)
Comment (juanvallejo) [reply tools/depcheck/README.md]: Sure, will link Origin example here

File: tools/depcheck/README.md (hunk 2 above)
Comment (juanvallejo) [reply tools/depcheck/README.md]: Thanks, will clarify this a bit more.
If copying a repository to a temporary location in order to expand symlinks, a GOPATH must be set to point to the new temporary location as well.

File: tools/depcheck/README.md (hunk 3 above)
Comment (juanvallejo) [reply tools/depcheck/README.md]: Thanks, will add a `Running` section

File: tools/depcheck/README.md (hunk 7 above)
Comment (juanvallejo) [reply tools/depcheck/README.md]: Thanks, will look into this. The `cmd/service-catalog` path should have been ignored as part of --openshift defaults

File: tools/depcheck/README.md (hunk 5 above)
Comment (juanvallejo) [reply tools/depcheck/README.md]: Thanks, will update overall structure of this doc

File: tools/depcheck/README.md (hunk 7 above)
Comment (soltysh) [reply tools/depcheck/README.md]: Yeah, but ensure there's an example how to include it, when service-catalog folks will want to use your tool as well. And most importantly that it also works for them.

Comment (soltysh) [reply tools/depcheck/README.md]: Although, you might sync with them if this requires additional eng effort, before actually doing it.

Comment (juanvallejo) [reply tools/depcheck/README.md]: This issue is addressed in the last commit. Although a user is able to --exclude a list of Go import paths, those are only excluded when building the graph, not when traversing the directory tree. The new patch makes it so that any errors encountered while traversing an "excluded" path are non-fatal.

> Yeah, but ensure there's an example how to include it, when service-catalog folks will want to use your tool as well. And most importantly that it also works for them.

If they wish to use this tool, they either have to run it against their own independent repo, or include all of their specific vendored deps in the openshift/origin vendor directory, as Go will not look in their own nested vendor tree for those deps.

File: tools/depcheck/README.md (hunk 4 above)
Comment (juanvallejo) [reply tools/depcheck/README.md]: Done

File: tools/depcheck/README.md (hunk 1 above)
Comment (juanvallejo) [reply tools/depcheck/README.md]: Done

**Comment by juanvallejo** [comment]
@soltysh thanks for the review, comments addressed.

File: tools/depcheck/README.md [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+```
+
+After this step, you should have a `depcheck` binary in your current working directory.
+
+
+<a name="running"></a>
+### Running
+
+The `depcheck` tool can be used to obtain `dot` output for a given repository.
+This output may be fed into tools like [dot (1)](https://www.systutorials.com/docs/linux/man/1-dot/)
```
Comment (soltysh) [inline tools/depcheck/README.md]: s/dot (1)/dot

File: tools/depcheck/README.md [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+The `depcheck` tool can further be used to `pin` dependencies from a `glide.lock` file onto a `glide.yaml` file.
+*Pinning* these dependencies means that you are taking all dependencies listed in a lock file that are not currently present
+in the given `glide.yaml` file, and generating a new `glide.yaml` file containing this new set of dependencies, each one pinned to
+the commit hash specified in the lock file.
+
+Example:
+
+```bash
+$ depcheck pin --glide.yaml=glide.yaml --glide.lock=glide.lock > glide.yaml
```
Comment (soltysh) [inline tools/depcheck/README.md]: Change the final `glide.yaml` to `newglide.yaml`, if you invoked above command you're get an empty glide.yaml file error.
